### PR TITLE
fix clang build on windows for zlib/1.2.11

### PR DESCRIPTION
**Pull request by AndreyMlashkin**
Specify library name and version:  **zlib/1.2.11@**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by AndreyMlashkin** [comment]
Fixes https://github.com/conan-io/conan-center-index/issues/9806

**Comment by SpaceIm** [comment]
What is the recommended profile when using clang-cl?

**Comment by ericLemanissier** [comment]
it does not. you should have compiler="Visual Studio" and compiler.toolset="ClangCL"

**Comment by SSE4** [comment]
> /cc @SSE4 I know that you have submitted several PRs to improve clang-cl support in conan, so maybe you know what should be the profile for this compiler.

I need to refresh my memory a bit, it was a while ago. I'll also check these `clang-cl` issues, if I can help moving em forward.
in short, there was no canonical way, but there are few different ways to do it:
1. `compiler=clang CC=clang.exe` - it uses GNU style command line options
2. `compiler=clang CC=clang-cl.exe` - it uses Visual C++ compatible command line options (it also might activate an additional compatibility with Microsoft C++ extension, like dllexport and so on)
3. `compiler=Visual Studio toolset=ClangCL` - very limited, doesn't really work for all the build systems (except probably CMake and MBuild

to complicate, there were also several ways to install clang, e.g.:
- binaries from llvmg.org
- one Microsoft supplies with Visual Studio (there were actually several of them, but C2 was abandoned long ago)
- `clang` package in MSYS - might not be fully compatible with previous two, because it uses a different `--target`

**Comment by AndreyMlashkin** [comment]
@SSE4, so this PR fixes the case number 2 from your list.
Could you please approve

**Comment by SSE4** [comment]
> @SSE4, so this PR fixes the case number 2 from your list. Could you please approve

I am looking. as it is under `if Windows`, it shouldn't break clang on Linux/Mac.
however, it might not work for clang installed via msys, I need to check that corner case first (and is it possible to distinguish them two at all).

**Comment by AndreyMlashkin** [comment]
good, thank you

**Review by SSE4 — APPROVED** [review]

File: recipes/zlib/all/conanfile.py [hunk 1]
```diff
@@ -101,14 +101,14 @@ def _rename_libraries(self):
             suffix = "d" if self.settings.build_type == "Debug" else ""
 
             if self.options.shared:
-                if self._is_msvc and suffix:
+                if (self._is_msvc or str(self.settings.compiler) in ("clang")) and suffix:
```
Comment (ericLemanissier) [inline recipes/zlib/all/conanfile.py:104]:
```suggestion
                if (self._is_msvc or self.settings.compiler == "clang") and suffix:
```

File: recipes/zlib/all/conanfile.py [hunk 2]
```diff
@@ -101,14 +101,14 @@ def _rename_libraries(self):
...
                     current_lib = os.path.join(lib_path, "zlib%s.lib" % suffix)
                     tools.rename(current_lib, os.path.join(lib_path, "zlib.lib"))
             else:
-                if self._is_msvc:
+                if self._is_msvc or str(self.settings.compiler) in ("clang"):
```
Comment (ericLemanissier) [inline recipes/zlib/all/conanfile.py:108]:
```suggestion
                if self._is_msvc or self.settings.compiler == "clang":
```

File: recipes/zlib/all/conanfile.py [hunk 3]
```diff
@@ -101,14 +101,14 @@ def _rename_libraries(self):
...
                     current_lib = os.path.join(lib_path, "zlibstatic%s.lib" % suffix)
                     tools.rename(current_lib, os.path.join(lib_path, "zlib.lib"))
-                elif self.settings.compiler in ("clang", "gcc", ):
+                elif self.settings.compiler in ("gcc", ):
```
Comment (ericLemanissier) [inline recipes/zlib/all/conanfile.py:111]:
```suggestion
                elif self.settings.compiler == "gcc":
```

**Comment by SpaceIm** [comment]
> but I don't understand why does it trigger CMake to choose a different file name

Is clang from msys2 a cl like compiler?

**Comment by SSE4** [comment]
> Is clang from msys2 a cl like compiler?

that's not exactly how it works. any clang has a `--driver-mode`. `clang.exe` defaults to `--driver-mode=gnu`, and `clang-cl.exe` defaults to `--driver-mode=cl`. they both (clang from MSYS and clang from Visual Studio) can understand GNU-style flags and CL-style flags just fine. that's just a matter of command line parser, and how do they interpret options.
that's really different between them is `target` (`i686-pc-windows-msvc` vs `x86_64-pc-windows-msys`). I think it might have an effect on code generation.

**Comment by SpaceIm** [comment]
> that's not exactly how it works. any clang has a `--driver-mode`. `clang.exe` defaults to `--driver-mode=gnu`, and `clang-cl.exe` defaults to `--driver-mode=cl`. they both (clang from MSYS and clang from Visual Studio) can understand GNU-style flags and CL-style flags just fine. that's just a matter of command line parser, and how do they interpret options.
> that's really different between them is `target` (`i686-pc-windows-msvc` vs `x86_64-pc-windows-msys`). I think it might have an effect on code generation.

I mean in CMake logic. `MSVC` variable in CMake refers to cl like compilers. So this PR must detect from profile settings when zlib CMakeLists will go in the `MSVC`branch.

**Comment by SSE4** [comment]
> I mean in CMake logic. `MSVC` variable in CMake refers to cl like compilers. So this PR must detect from profile settings when zlib CMakeLists will go in the `MSVC`branch.

actually, it seems like it has nothing to do with compiler 🤔 
the logic to select output name is https://github.com/madler/zlib/blob/v1.2.11/CMakeLists.txt#L204

outputs:
```
cmake .. -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ 
-- The C compiler identification is Clang 13.0.0 with GNU-like command-line
-- The CXX compiler identification is Clang 13.0.0 with GNU-like command-line
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/bin/clang.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/Llvm/bin/clang++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- WIN32 = TRUE
-- MSVC = FALSE
-- UNIX = FALSE
-- Configuring done
-- Generating done
-- Build files have been written to: C:/conan/conan-center-index/recipes/test/build
```
vs
```
 cmake .. -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -G Ninja
-- The C compiler identification is Clang 11.0.0
-- The CXX compiler identification is Clang 11.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- WIN32 = FALSE
-- MSVC = FALSE
-- UNIX = TRUE
-- Configuring done
-- Generating done
-- Build files have been written to: /c/conan/conan-center-index/recipes/test/build
```
so it's CMake in MSYS spoiling all the things.

**Comment by SSE4** [comment]
> @SSE4 there is a compiler logic in addition to the one you've highlighted: https://github.com/madler/zlib/blob/v1.2.11/CMakeLists.txt#L62-L67
> 
> Moreover, implicitly PREFIX value is compiler dependent.

fair, I didn't encounter it as I was testing `build_type = Release` only.

anyway, I suppose it's MSYS CMake difference, and at this point, I am not sure how to properly fix the recipe, so the library name is predictable on Windows. that I can say for sure, it has nothing to do with `clang` flavor.

the same will happen with MSYS/Cygwin GCC (MinGW), compare to running MinGW from cmd.exe. it's because CMake in MSYS  doesn't assume it's running on WIN32, but on UNIX.

**Comment by AndreyMlashkin** [comment]
What's wrong with the build? I don't see error log
@jgsogo

**Comment by jgsogo** [comment]
The CI is failing at the very beginning (even before starting) with:

```
Started by user [admin](https://ci-conan-prod.jfrog.team/user/admin)
ERROR: No flow definition, cannot run
Finished: FAILURE
```

I'll try closing and opening this PR

**Comment by jgsogo** [comment]
Ok, now at least it has started... maybe it was related to the failures that Github experienced yesterday.

**Review by ericLemanissier — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Pull request closed without merging**